File: haskell/bitcoin/Number/Bench_Number.hs

-- This code is running fine, but unlikely to be measuring the performance
-- of anything. The lesson is: writing bench-marking code in haskell is hard.

import Bench_Abstract
import Rand
import Number

main :: IO ()
main = toIO run

run :: Rand ()
run = do
  logMessage "Number benchmark running ..."

  let all = True

  benchFromBytes
  benchToBytes
  benchAdd
  benchMul
  benchShow
  benchRandom

  if all then do
    liftIO benchZero
    benchOne
    benchSign
    benchCompareTo
    benchHash
    benchNumberEquals
    benchFromInteger
    benchToInteger
    benchBitLength
  else
    return ()

benchZero :: IO ()
benchZero = do 
  benchmark (return (zero :: Number)) "zero" 1000000
  benchmark (return (0 :: Integer)) "zero*" 1000000


benchOne :: Rand ()
benchOne = do
  benchmark (return (one :: Number)) "one" 1000000
  benchmark (return (1 :: Integer)) "one*" 1000000


benchFromBytes :: Rand ()
benchFromBytes = do
  bytes <- getRandomBytes 32
  benchmark ( do 
    fromBytes (Sign 1)  bytes :: Rand Number
    fromBytes (Sign (-1)) bytes :: Rand Number
    ) "fromBytes" 1000000

benchToBytes :: Rand ()
benchToBytes = do
  x <- random (NumBits 256) :: Rand Number
  let y = negate x
  let n = toInteger x
  let m = toInteger y 
  benchmark ( do
    toBytes x (NumBytes 32)
    toBytes y (NumBytes 32)
    ) "toBytes" 1000000



benchAdd :: Rand()
benchAdd = do
  x <- random (NumBits 256) :: Rand Number
  y <- negate <$> random (NumBits 256) :: Rand Number
  let n = toInteger x
  let m = toInteger y
  benchmark ((return $ x + y) >> (return $ y + x)) "add" 1000000
  benchmark ((return $ n + m) >> (return $ m + n)) "add*" 1000000

benchMul :: Rand()
benchMul = do 
  x <- random (NumBits 256) :: Rand Number
  y <- negate <$> random (NumBits 256) :: Rand Number
  let n = toInteger x
  let m = toInteger y
  benchmark ((return $ x * y) >> (return $ y * x)) "mul" 1000000
  benchmark ((return $ n * m) >> (return $ m * n)) "mul*" 1000000

  

benchShow :: Rand()
benchShow = do
  x <- random (NumBits 256) :: Rand Number
  y <- negate <$> random (NumBits 256) :: Rand Number
  let n = toInteger x
  let m = toInteger y
  benchmark ((return $ show x) >> (return $ show y)) "show (10k)" 10000
  benchmark ((return $ show n) >> (return $ show m)) "show* (10k)" 10000


benchRandom :: Rand()
benchRandom = benchmark (random (NumBits 256) :: Rand Number) "random (10k)" 10000

benchSign :: Rand()
benchSign = do
  x <- random (NumBits 256) :: Rand Number
  y <- negate <$> random (NumBits 256) :: Rand Number
  let n = toInteger x
  let m = toInteger y
  benchmark ((return $ sign x) >> (return $ sign y)) "sign" 1000000
  benchmark ((return $ signum x) >> (return $ signum y)) "signum" 1000000
  benchmark ((return $ signum n) >> (return $ signum m)) "signum*" 1000000


benchCompareTo :: Rand()
benchCompareTo = do
  x <- random (NumBits 256) :: Rand Number
  y <- random (NumBits 256) :: Rand Number
  let n = toInteger x
  let m = toInteger y
  benchmark ((return $ x <= y) >> (return $ y <= x)) "<=" 1000000
  benchmark ((return $ x < y) >> (return $ y < x)) "<" 1000000
  benchmark ((return $ n <= m) >> (return $ m <= n)) "<=*" 1000000
  benchmark ((return $ n < m) >> (return $ m < n)) "<*" 1000000



benchHash :: Rand()
benchHash = do
  x <- random (NumBits 256) :: Rand Number
  let n = toInteger x
  benchmark (return $ hash x) "hash" 1000000
  benchmark (return $ hash n) "hash*" 1000000
 

benchNumberEquals :: Rand()
benchNumberEquals = do
  x <- random (NumBits 256) :: Rand Number
  y <- random (NumBits 256) :: Rand Number
  let n = toInteger x
  let m = toInteger y
  benchmark ((return $ x == y) >> (return $ y == x)) "==" 1000000
  benchmark ((return $ n == m) >> (return $ m == n)) "==*" 1000000



benchFromInteger :: Rand()
benchFromInteger = do
  x <- random (NumBits 256) :: Rand Number
  let y = negate x
  let n = toInteger x
  let m = toInteger y
  benchmark ((return $ fromInteger n) >> (return $ fromInteger m)) 
            "fromInteger" 1000000

benchToInteger :: Rand()
benchToInteger = do
  x <- random (NumBits 256) :: Rand Number
  let y = negate x
  benchmark ((return $ toInteger x) >> (return $ toInteger y))
            "toInteger" 1000000


benchBitLength :: Rand()
benchBitLength = do
  x <- random (NumBits 256) :: Rand Number
  let n = toInteger x
  benchmark (return $ bitLength x) "bitLength" 1000000

File: python/bitcoin/Number/Bench_Number.py
import Number
from Bench_Abstract import *

class Bench_Number(Bench_Abstract):

    def run(self):
        logMessage("Number benchmark running ...")

        runAll = True

        benchFromBytes()
        benchToBytes()
        benchAdd()
        benchMul()
        benchNegate()
        benchToString()
        benchRandom()

        if runAll:
            benchZERO()
            benchONE()
            benchSign()
            benchCompareTo()
            benchHashCode()
            benchNumberEquals()  
            benchFromBigInteger()
            benchToInt()
            benchBitLength()
 

def benchZERO():
    def test1():
        x = Number.ZERO
    def test2():
        x = 0
    benchmark(test1, "ZERO", 1000000)
    benchmark(test2, "ZERO*", 1000000)

def benchONE():
    def test1():
        x = Number.ONE
    def test2():
        x = 1
    benchmark(test1, "ONE", 1000000)
    benchmark(test2, "ONE*", 1000000)

def benchFromBytes():
    bs = getRandomBytes(32)
    def test1():
        Number.fromBytes(1, bs)
        Number.fromBytes(-1, bs)
    def test2():
        int.from_bytes(bs, byteorder='big', signed=False)
        -int.from_bytes(bs, byteorder='big', signed=False)

    benchmark(test1, "fromBytes (10k)", 10000)  # 1 million too slow
    benchmark(test2, "fromBytes* (10k)", 10000) # 1 million too slow


def benchToBytes():
    x = Number.random(256)
    y = -x
    n = int(x)
    m = int(y)
    def test1():
        x.toBytes(32)
        y.toBytes(32)
    def test2():
        n.to_bytes(32, byteorder='big', signed=False)
        m.to_bytes(33, byteorder='big', signed=True)
    benchmark(test1, "toBytes (10k)", 10000)    # 1 million too slow
    benchmark(test2, "toBytes* (10k)", 10000)   # 1 million too slow

def benchAdd():
    x = Number.random(256)
    y = -Number.random(256)
    n = int(x)
    m = int(y)
    def test1():
        x + y
        y + x
    def test2():
        n + m
        m + n
    benchmark(test1, "add (10k)", 10000)
    benchmark(test2, "add* (10k)", 10000)

def benchMul():
    x = Number.random(256)
    y = -Number.random(256)
    n = int(x)
    m = int(y)
    def test1():
        x * y
        y * x
    def test2():
        n * m
        m * n
    benchmark(test1, "mul (10k)", 10000)
    benchmark(test2, "mul* (10k)", 10000)

def benchNegate():
    x =  Number.random(256)
    y = -Number.random(256)
    n = int(x)
    m = int(y)
    def test1():
        -x
        -y
    def test2():
        -n
        -m
    benchmark(test1, "negate (10k)", 10000)
    benchmark(test2, "negate* (10k)", 10000)

def benchToString():
    x = Number.random(256)
    y = -x
    n = int(x)
    m = int(y)
    def test1():
        str(x)
        str(y)
    def test2():
        str(n)
        str(m)
    benchmark(test1, "str (10k)", 10000)
    benchmark(test2, "str* (10k)", 10000)



def benchRandom():
    benchmark(lambda : Number.random(256), "random (10k)", 10000)


def benchSign():
    x = Number.random(256)
    y = -x
    n = int(x)
    m = int(y)
    def test1():
        x.sign()
        y.sign()
    benchmark(test1, "sign", 1000000)


def benchCompareTo():
    x = Number.random(256)
    y = Number.random(256)
    n = int(x)
    m = int(y)
    def test1():
        x < y
        y < x
    def test2():
        n < m
        m < n
    benchmark(test1, "compareTo", 1000000)
    benchmark(test2, "compareTo*", 1000000)

def benchHashCode():
    x = Number.random(256)
    n = int(x)
    benchmark(lambda : hash(x), "hash", 1000000)
    benchmark(lambda : hash(n), "hash*", 1000000)

def benchNumberEquals():
    x = Number.random(256)
    y = Number.random(256)
    n = int(x)
    m = int(y)
    def test1():
        x == y
        y == x
    def test2():
        n == m
        m == n
    benchmark(test1, "equals", 1000000)
    benchmark(test2, "equals*", 1000000)


def benchFromBigInteger():
    x = Number.random(256)
    n = int(x)
    m = -n
    def test1():
        Number.fromInt(n)
        Number.fromInt(m)
    benchmark(test1, "fromInt", 1000000)

def benchToInt():
    x = Number.random(256)
    y = -x
    def test1():
        int(x)
        int(y)
    benchmark(test1, "toInt", 1000000)

def benchBitLength():
    x = Number.random(256)
    n = int(x)
    benchmark(lambda : x.bitLength(), "bitLength", 1000000)
    benchmark(lambda : n.bit_length(), "bitLength*", 1000000)


bench = Bench_Number()
bench.run()




